many: configure store from state, reconfigure store at runtime

atomatt opened the pull request:
Configures the store's API from the state - recorded as a new `"store": {"api": "<url>"}` section in the state's data. Also adds a func to reconfigure the store to use a new base API URL at runtime.

The cached Store was previously managed by snapstate but that's moved to a new storestate package, alongside other store-related state helpers.

Note: caching the auth context in the state is a bit of a hack, but is the only sane way I could find (i.e. without affecting lots of code & tests) to make it available so the store's API can be changed by a call to a storestate func instead of going through the overlord.

Depends on https://github.com/snapcore/snapd/pull/3580.

chipaca commented:
As this has not been reviewed yet, and is a mess of merges, could you rebase it now that its prerequisite is merged? Reviewing is going to be a nightmare otherwise

chipaca commented:
Is there a reason this adds a new endpoint instead of using config?

atomatt commented:
@chipaca rebase on its way soon, but what did you mean by "new endpoint instead of using config"?

chipaca commented:
@atomatt the thing that you're doing here looks, to me, to be no different from for example the refresh schedule, which is handled via `POST`ing to `/v2/snaps/core/conf` (“`snap set core refresh.schedule yadda`”)

chipaca commented:
(and that comment might've been meant for #3586...)

atomatt commented:
@chipaca yeah, I figured that out, but it does have some bearing on this branch too I think.

Firstly, I *really* like the idea of using `snap set core ...` for this, so thanks for the idea! I think it would look something like the following:

- allow `snap set core custom-store ...` to write config, but leave the state engine to pick up the change
- do something similar to snapmgr's Ensure/ensureRefreshes to detect the config change
- attempt to find assertion, grab address, etc
- replace Store in the state's cache

Does that make sense so far?

I spent some time poking around in snapd to see how viable this is and there are a couple of things that may be an issue.

I *think* we only want root to be able to change the store's API but `snap set` is available to any authenticated user. (It seems you have to be root to `snap login` anyway, so you were root at some point, but that feels too implicit for a security check.)

Also, if the core snap is not yet installed then there's nothing to `snap set core ...` against and it fails. This prevents the main scenario we're trying to address, e.g. a classic machine on a constrained network that *requires* a custom store API for an edge proxy with access to an upstream store containing core and other snaps. Maybe there's an argument that `snap set` for core could be special-cased to work before it's actually installed since there's such a close relationship between snapd and core?

atomatt commented:
Rebased and cleaned up.

@chipaca would appreciate your thoughts about problems using `snap set core ...` instead of this and the daemon/cli branches.

sparkiegeek commented:
I rebased this in #3780 (Matt is on holiday)
